### ISSUE TITLE
[HOLD for payment 2024-05-06] [$250] [Held requests] Hold request link for the requestee shows "undefined"

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#37554]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.45-2
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Go to staging.new.expensify.com
2. [User A] Request money from User B
3. [User B] Navigate to request details page
4. [User B] Click 3-dot menu > Hold request

## Expected Result:
There will be no "undefined" in Hold request link

## Actual Result:
Hold request link for the requestee shows "undefined"

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/0929ef75-884b-4eb2-af21-d641aca6ce9d

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0166aa9b176af4b313</li>
        <li>Upwork Job ID: 1766289313058312192</li>
        <li>Last Price Increase: 2024-04-06</li>
    </ul>
</details>

**Comment by lanitochka17** [comment]
@greg-schroeder  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by Pujan92** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Hold request reason link includes undefined 

### What is the root cause of that problem?
We are passing policy type here instead of the iouType and for a non-active policy it will be undefined.
https://github.com/Expensify/App/blob/bae502e031a4fc9c4925636f2a05c70297300c6e/src/components/MoneyRequestHeader.tsx#L98
https://github.com/Expensify/App/blob/bae502e031a4fc9c4925636f2a05c70297300c6e/src/ROUTES.ts#L301-L304

### What changes do you think we should make in order to solve the problem?
Pass `CONST.IOU.TYPE.REQUEST` for the iouType to make the route consistent with other routes.

**Comment by greg-schroeder** [comment]
[Checking w/ the team on this one](https://expensify.slack.com/archives/C01SKUP7QR0/p1709340289530589)

**Comment by greg-schroeder** [comment]
Setting to External per team discussion

**Comment by damadoko** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Hold request reason link includes undefined
### What is the root cause of that problem?
- The `policyID` of a `Chat report` is created by the requestor's personal `policyID` (in the BE side).
- Each account only has 1 personal `policyID`. That mean when the receiver open above report, the report `policyID` will not found in receiver's policy list, it will fallback to `{}`:https://github.com/Expensify/App/blob/9302ade9988b1cba3e07b48f474bc35a6deea7fb/src/pages/home/ReportScreen.js#L280
- That empty `policy` was passed to the navigate param, which cause the `undefined` value included in the link:https://github.com/Expensify/App/blob/9302ade9988b1cba3e07b48f474bc35a6deea7fb/src/components/MoneyRequestHeader.tsx#L90

### What changes do you think we should make in order to solve the problem?
I think we need to pass `CONST.POLICY.TYPE.PERSONAL` as the fallback value:
```js
 Navigation.navigate(ROUTES.MONEY_REQUEST_HOLD_REASON.getRoute(policy?.type ?? CONST.POLICY.TYPE.PERSONAL, parentReportAction?.originalMessage?.IOUTransactionID ?? '', report?.reportID, activeRoute));
```
Because currently, we only have [4 type of policies](https://github.com/Expensify/App/blob/9302ade9988b1cba3e07b48f474bc35a6deea7fb/src/CONST.ts#L1367-L1376), other type `free`, `corporate`, `team` are workspace's policy. Therefore, these policy type always available in both requester and receiver. This bug only happen in receiver side in P2P request money flow.

### What alternative solutions did you explore? (Optional)
N/A

**Comment by bernhardoj** [comment]
It shows not found page since this TS [PR](https://github.com/Expensify/App/pull/35566) where it puts an empty string as the default value. There is an issue for that [here](https://github.com/Expensify/App/issues/38345), but basically they are the same issue.

**Comment by sobitneupane** [comment]
@bernhardoj Do you mean https://github.com/Expensify/App/issues/38345 issue and https://github.com/Expensify/App/issues/37554 issue are same?

**Comment by bernhardoj** [comment]
I would say they are the same and @Pujan92 solution will solve it.

**Comment by sobitneupane** [comment]
Just for clarification:

@greg-schroeder Can you please clarify the expected output? We want to show the HoldRequest option for both requestee and requestor, right?

Test steps in https://github.com/Expensify/App/pull/33897 PR says all members should be able to put the request to hold in personal IOU.

> In top right corner click three dots and "Hold request" button (in a workspace people able to put the request on hold are: admins, approvers that the request has been sent to, and owner of the money request, in personal IOU all members should be able to put the request on hold)

**Comment by sobitneupane** [comment]
Thanks for the proposal @Pujan92. Can you please add more details in RCA

> We are passing policy type here instead of the iouType and for a non-active policy it will be undefined.

**Comment by Pujan92** [comment]
> Thanks for the proposal @Pujan92. Can you please add more details in RCA

@sobitneupane I updated my [proposal](https://github.com/Expensify/App/issues/37554#issuecomment-1971804128) to mention why it is showing now not-found page instead of undefined in the route which is also correctly pointed by @bernhardoj [here](https://github.com/Expensify/App/issues/37554#issuecomment-1998942109). The solution remains same to pass correct iouType.

**Comment by sobitneupane** [comment]
@greg-schroeder We are waiting for the clarification on the expected output https://github.com/Expensify/App/issues/37554#issuecomment-2003225695

**Comment by greg-schroeder** [comment]
Pretty sure this is under #wave-collect as it's re: hold request - I'm adding it accordingly. @trjExpensify or @JmillsExpensify  could you potentially weigh in on https://github.com/Expensify/App/issues/37554#issuecomment-2003225695 about the desired output here?

**Comment by trjExpensify** [comment]
To answer the question, the requestor should be able to hold their own requests on iouReports as well.

**Comment by greg-schroeder** [comment]
Not sure - @robertjchen @BartoszGrajdek do you think your recent work may have affected what's outlined here?

**Comment by robertjchen** [comment]
hrm, there hasn't been any backend changes that would change this behavior. It would appear that the determination for the error is made by frontend logic? https://github.com/Expensify/App/blob/ebc8de8229fa700e5a7d08f4589db934a79d4f1d/src/libs/ReportUtils.ts#L2345

Could we debug the logic there to see if that needs to be updated to match the requirements laid out in https://github.com/Expensify/App/issues/37554#issuecomment-2003225695 and if any backend changes are necessary?

**Comment by sobitneupane** [comment]
@robertjchen The main issue here is app crashes when the person who is requested tries to hold request (Issue only occurs in personal IOU). The person who is requesting can hold the request without any issue.

The main reason for the issue is in `MoneyRequestHeader` we use following piece of code to navigate to Hold Request page. But `policy` is null in this case(probably because this is personal IOU).

https://github.com/Expensify/App/blob/ebc8de8229fa700e5a7d08f4589db934a79d4f1d/src/components/MoneyRequestHeader.tsx#L100

I think we can fallback to `CONST.POLICY.TYPE.PERSONAL` as suggested [here](https://github.com/Expensify/App/issues/37554#issuecomment-1986819097)

But the person who is requested won't be able to hold the request which I think is the expected behavior.

**Comment by trjExpensify** [comment]
I got this one today btw: 

<img width="1507" alt="image" src="https://github.com/Expensify/App/assets/16232057/8956d3e5-7412-44a4-a162-236a2056f235">

**Comment by robertjchen** [comment]
I see- falling back to `CONST.POLICY.TYPE.PERSONAL` would be the right step there 👍 

> But the person who is requested won't be able to hold the request which I think is the expected behavior.

I think the intention is that they would be able to hold incoming requests, since "in personal IOU all members should be able to put the request on hold)"

**Comment by sobitneupane** [comment]
@trjExpensify 

https://github.com/Expensify/App/issues/37554#issuecomment-2034547766

This looks like a different issue. It was not reproducible earlier. 

So, It appears on the requestor side and the app crashes after the requestor holds the request.

Do we want to create a separate issue for [this](https://github.com/Expensify/App/issues/37554#issuecomment-2034547766) or do we want to handle it here? It has different root cause than this issue.

**Comment by sobitneupane** [comment]
> I think the intention is that they would be able to hold incoming requests, since "in personal IOU all members should be able to put the request on hold)"

@robertjchen As you pointed out [earlier](https://github.com/Expensify/App/issues/37554#issuecomment-2034344479), we are using [`canEditMoneyRequest`](https://github.com/Expensify/App/blob/ebc8de8229fa700e5a7d08f4589db934a79d4f1d/src/libs/ReportUtils.ts#L2345) to determine if the user can hold the request. And only the requestor has access to edit IOU request. So, the other user is not allowed to hold request.

https://github.com/Expensify/App/blob/ebc8de8229fa700e5a7d08f4589db934a79d4f1d/src/libs/ReportUtils.ts#L2375-L2377

**Comment by robertjchen** [comment]
Got it, can we make an exception there for hold-related actions?

**Comment by robertjchen** [comment]
re: the crash  https://github.com/Expensify/App/issues/37554#issuecomment-2036443707

It is unrelated to this and we have a backend fix for that coming out later today 👍 https://github.com/Expensify/App/issues/39428

**Comment by sobitneupane** [comment]
> Got it, can we make an exception there for hold-related actions?


I believe we can. 

@damadoko @Pujan92 @GandalfGwaihir Do you want to update your proposals to allow the person being requested to hold IOU request?

**Comment by sobitneupane** [comment]
Thanks for the update @GandalfGwaihir 

We are using a fallback value for policy type so, rather than `CONST.IOU.TYPE.REQUEST`, we should use `CONST.POLICY.TYPE.PERSONAL`.

If I am not wrong, `parentReportAction?.actionName` is 'IOU' even in the case of free, control and collect policy. The proposed change will always bypass the `canEditMoneyRequest` test. We want to allow  all members to hold request only in personal IOU.

> I think the intention is that they would be able to hold incoming requests, since "in personal IOU all members should be able to put the request on hold)"

**Comment by sobitneupane** [comment]
@GandalfGwaihir We are already passing the policy type as param:

https://github.com/Expensify/App/blob/ce003b2ac3d0f5c407c6da38ec673011032e8395/src/components/MoneyRequestHeader.tsx#L100

I believe we can use it to determine whether the IOU in concern belongs to personal policy type or not.

**Comment by sobitneupane** [comment]
@GandalfGwaihir Will you please update your proposal to reflect this [suggestion](https://github.com/Expensify/App/issues/37554#issuecomment-2042532316)?

**Comment by greg-schroeder** [comment]
Proposal review continues!

**Comment by sobitneupane** [comment]
[Proposal](https://github.com/Expensify/App/issues/37554#issuecomment-2003224729) from @GandalfGwaihir looks good to me.

🎀 👀 🎀 C+ reviewed

**Comment by dangrous** [comment]
Hi! I think this sounds good, but I'm going to take a little more time to confirm, since I just came back from leave and the Hold Request flow didn't exist when I left - so I have to get more familiar with it!

**Comment by greg-schroeder** [comment]
Thanks @dangrous! Let us know if you need more eyes 👍

**Comment by dangrous** [comment]
Okay, I think this makes sense to me! Thanks for your patience. I may have a couple questions along the way but let's move forward.

**Comment by dangrous** [comment]
Yep! I was out, looking at this shortly.

**Comment by sobitneupane** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
> 
> * [ ]  [@sobitneupane] The PR that introduced the bug has been identified. Link to the PR:

https://github.com/Expensify/App/pull/33897
> * [ ]  [@sobitneupane] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

https://github.com/Expensify/App/pull/33897#issuecomment-2097942111
> * [ ]  [@sobitneupane] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
> * [ ]  [@sobitneupane] Determine if we should create a regression test for this bug.

Yes
> * [ ]  [@sobitneupane] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

**Comment by sobitneupane** [comment]
**Regression Test Proposal**

1. Login from one account
2. Request money 1:1 with the same user 2 times
3. Login from that other account to which you have requested money
4. Open chat and navigate to any one of the money request by the previously logged in user.
5. Hold the request
6. Verify that you can successfully hold the request.

Do we agree 👍 or 👎

**Comment by greg-schroeder** [comment]
Processing

**Comment by greg-schroeder** [comment]
$250 - C+ - @sobitneupane - you can make a ND manual request for payment
$250 - C - @GandalfGwaihir  - I've sent you a manual offer for this issue, can you please accept it in upwork? Thanks!

I'll now file the regression test.

**Comment by JmillsExpensify** [comment]
$250 approved for @sobitneupane
